core/src/main/kotlin/com/malinskiy/marathon/execution/device/DeviceActor.kt

```diff
@@ -137,25 +143,43 @@ class DeviceActor(private val devicePoolId: DevicePoolId,
         }
     }
 
-    private fun sendResults(result: CompletableDeferred<TestBatchResults>) {
-        launch {
-            val testResults = result.await()
+    private fun sendResults(testBatch: TestBatch?, results: CompletableDeferred<TestBatchResults>?): Job = launch {
+        if (results == null) {
+            returnBatch(testBatch)
+            return@launch
+        }
+
+        val innerJob = launch {
+            logger.debug("Awaiting batch results to DevicePool: ${device.serialNumber}")
+            val testResults = results.await()
             pool.send(DevicePoolMessage.FromDevice.CompletedTestBatch(device, testResults))
+            logger.debug("Sent batch results to DevicePool: ${device.serialNumber}")
+        }
+        delay(AWAIT_BATCH_RESULTS_TIMEOUT_MS)
```

Comment:
oi vey. this _doesn't_ look good. could you please reiterate what _exactly_ is this timeout supposed to help with?

Reply:
Ohh, that's is simple. It helps to avoid the job hanging on await command indefinitely. So if something happens and we need to send results, we expect them to appear on Deferred in the next 500ms. if that never happened, we will cancel the coroutine and return the batch.

During my experiments, I actually have seen devices hanging with their job blocked by await command.

Reply:
FYI: https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/with-timeout.html

Reply:
Thank you Anton, that's exactly what I was looking for :)